source/server/listener_manager_impl.cc

```diff
@@ -242,11 +230,101 @@ ListenerImpl::~ListenerImpl() {
+  // Save mappings.
+  if (server_names.empty()) {
+    addFilterChainForServerName(EMPTY_STRING, transport_socket_name, filter_chain);
+  } else {
+    for (const auto& server_name : server_names) {
+      if (isWildcardServerName(server_name)) {
+        // Add mapping for the wildcard domain, i.e. ".example.com" for "*.example.com".
+        addFilterChainForServerName(server_name.substr(1), transport_socket_name, filter_chain);
+      } else {
+        addFilterChainForServerName(server_name, transport_socket_name, filter_chain);
+      }
+    }
+  }
+}
+
+void ListenerImpl::addFilterChainForServerName(const std::string& server_name,
+                                               const std::string& transport_socket_name,
+                                               Network::FilterChainSharedPtr filter_chain) {
+  auto& server_name_match = filter_chains_[server_name];
+  server_name_match[transport_socket_name] = filter_chain;
+
+  // Add filter chain mapping for an empty transport socket name, so that it matches also when
+  // no transport protocol match is configured in the filter chain match.
+  if (!transport_socket_name.empty() &&
+      server_name_match.find(EMPTY_STRING) == server_name_match.end()) {
+    server_name_match[EMPTY_STRING] = filter_chain;
+  }
+}
+
+const Network::FilterChainSharedPtr
+ListenerImpl::findFilterChain(Network::ConnectionSocketPtr& socket) const {
+  const std::string server_name(socket->requestedServerName());
+
+  // Match on exact server name, i.e. "www.example.com" for "www.example.com".
+  const auto server_name_exact_match = filter_chains_.find(server_name);
+  if (server_name_exact_match != filter_chains_.end()) {
+    return findFilterChainForServerName(server_name_exact_match->second, socket);
+  }
+
+  // Match on the wildcard domain, i.e. ".example.com" for "www.example.com".
+  const size_t pos = server_name.find('.');
+  if (pos > 0 && pos < server_name.size() - 1) {
+    const std::string wildcard = server_name.substr(pos);
+    const auto server_name_wildcard_match = filter_chains_.find(wildcard);
+    if (server_name_wildcard_match != filter_chains_.end()) {
+      return findFilterChainForServerName(server_name_wildcard_match->second, socket);
+    }
+  }
+
+  // Catch-all for any server name.
+  const auto server_name_catchall_match = filter_chains_.find(EMPTY_STRING);
+  if (server_name_catchall_match != filter_chains_.end()) {
+    return findFilterChainForServerName(server_name_catchall_match->second, socket);
+  }
+
+  return nullptr;
+}
+
+const Network::FilterChainSharedPtr ListenerImpl::findFilterChainForServerName(
+    const std::unordered_map<std::string, Network::FilterChainSharedPtr>& server_name_match,
+    Network::ConnectionSocketPtr& socket) const {
+  const std::string transport_protocol_name(socket->detectedTransportProtocol());
+
+  // Match on exact transport protocol, e.g. "tls".
```

Comment:
I think it would be great to document in user facing documentation somewhere what the rules regarding order are here.

Reply:
Agreed, I'll write something up tomorrow.